Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp

```diff
@@ -0,0 +1,152 @@
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+
+#include "ConsoleMessageTest.h"
+#include "WebKitTestServer.h"
+
+static const char indexHTML[] =
+"<!DOCTYPE html><html><body>"
+"<button id='enterXR'>Enter Experience</button>"
+"<script>"
+"document.getElementById('enterXR').addEventListener('click', () => {"
+"  navigator.xr.requestSession('immersive-vr').then(session => {"
+"    console.log('XR session started');"
+"    session.addEventListener('end', (event) => {"
+"        console.log('XR session ended');"
+"    });"
+"  }).catch(err => console.error(`XR session failed to start: ${err}`));"
+"});"
+"</script></body></html>";
+
+class ImmersiveModeTest : public ConsoleMessageTest {
+public:
+    MAKE_GLIB_TEST_FIXTURE(ImmersiveModeTest);
+
+    static void isImmersiveModeEnabledChanged(GObject*, GParamSpec*, ImmersiveModeTest* test)
+    {
+        g_signal_handlers_disconnect_by_func(test->webView(), reinterpret_cast<void*>(isImmersiveModeEnabledChanged), test);
+        g_main_loop_quit(test->m_mainLoop);
+    }
+
+    void waitUntilIsImmersiveModeEnabledChanged()
+    {
+        g_signal_connect(m_webView.get(), "notify::is-immersive-mode-enabled", G_CALLBACK(isImmersiveModeEnabledChanged), this);
+        g_main_loop_run(m_mainLoop);
+    }
+
+    bool isImmersiveModeEnabled() const
+    {
+        return webkit_web_view_is_immersive_mode_enabled(m_webView.get());
+    }
+
+    void leaveImmersiveMode()
+    {
+        webkit_web_view_leave_immersive_mode(m_webView.get());
+    }
+
+    struct NetworkPolicyGuard {
+        NetworkPolicyGuard(ImmersiveModeTest* test, WebKitTLSErrorsPolicy policy)
+            : m_test(test)
+        {
+            // Ignore server certificate errors.
+#if ENABLE(2022_GLIB_API)
+            m_originalPolicy = webkit_network_session_get_tls_errors_policy(m_test->m_networkSession.get());
+            webkit_network_session_set_tls_errors_policy(m_test->m_networkSession.get(), policy);
+#else
+            m_manager = webkit_web_context_get_website_data_manager(m_test->m_webContext.get());
+            m_originalPolicy = webkit_website_data_manager_get_tls_errors_policy(m_manager);
+            webkit_website_data_manager_set_tls_errors_policy(m_manager, policy);
+#endif
+        }
+        ~NetworkPolicyGuard()
+        {
+#if ENABLE(2022_GLIB_API)
+            webkit_network_session_set_tls_errors_policy(m_test->m_networkSession.get(), m_originalPolicy);
+#else
+            webkit_website_data_manager_set_tls_errors_policy(m_manager, m_originalPolicy);
+#endif
+        }
+        ImmersiveModeTest* m_test;
+        WebKitTLSErrorsPolicy m_originalPolicy;
+#if !ENABLE(2022_GLIB_API)
+        WebKitWebsiteDataManager* m_manager;
+#endif
+    };
```

Comment:
This is a great idea, if you move this to WebViewTest it could be used from TestSSL.cpp